### PR TITLE
Elm 130 aggregate locations

### PR DESCRIPTION
**Pull request by Dtzz81**
We added LocationsAggregation - delivering location data in aggregative format, added duration (with default option 1hr, max 24 hours). Added tests for ISO DateTime formats (bad request, status ok) and fixed issues with date time format inputs. 

https://dsdmoj.atlassian.net/browse/ELM-130?atlOrigin=eyJpIjoiZDIzNzQ2Y2YxZDBjNDE5NzliNDQ5NWJhM2M5NDI2N2UiLCJwIjoiaiJ9

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppselectronicmonitoringdataplatformapi/model/LocationAggregation.kt [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+import java.util.*
+
+interface LocationAggregation {
+  var latitude: Double
+  var longitude: Double
+  var datetime: Date
+  fun setValue(latitude: Double, longitude: Double, datetime: Date)
+}
+
+class LocationAggregationDemo(override var latitude: Double, override var longitude: Double, override var datetime: Date) : LocationAggregation {
```
Comment (JazJax) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppselectronicmonitoringdataplatformapi/model/LocationAggregation.kt:12]: Is there a reason this is called 'demo'? We should probably rename this to imply it's functional, even if functionality has limits.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppselectronicmonitoringdataplatformapi/controller/DeviceWearerControllerTest.kt [hunk 2]
```diff
@@ -41,12 +41,11 @@ class DeviceWearerControllerTest {
 
   @Test
   fun `getAllDeviceWearers should return Ok with a message when there are no device wearers`() {
-    Mockito.`when`(deviceWearerService.getAllDeviceWearers()).thenReturn(listOf<DeviceWearer>())
```
Comment (JazJax) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppselectronicmonitoringdataplatformapi/controller/DeviceWearerControllerTest.kt:44]: @PedroSoaresMoj I think this was a you change - can you comment on the reason? Seems like it makes this test do nothing, when combined with the change on old-line 49.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppselectronicmonitoringdataplatformapi/controller/DeviceWearerControllerTest.kt [hunk 3]
```diff
@@ -69,11 +68,10 @@ class DeviceWearerControllerTest {
   fun `getDeviceWearerByID should return Ok with a message when device wearer does not exist`() {
     val id = "c6a5aa13-c948-41ca-8d6a-dfe9ff2e8fd3"
 
-    val response = null
-    Mockito.`when`(deviceWearerService.getDeviceWearerById(id)).thenReturn(response)
```
Comment (JazJax) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppselectronicmonitoringdataplatformapi/controller/DeviceWearerControllerTest.kt:73]: @PedroSoaresMoj as above comment

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppselectronicmonitoringdataplatformapi/controller/LocationControllerTest.kt [hunk 4]
```diff
@@ -33,10 +36,17 @@ class LocationControllerTest {
...
+    expected: ResponseEntity<LocationAggregationResponse>,
+  ) {
+    Assertions.assertThat(result.body?.error).isEqualTo("")
+    Assertions.assertThat(result.body?.message).isEqualTo(expected.body?.message)
+    Assertions.assertThat(result.statusCode).isEqualTo(HttpStatus.OK)
+  }
+
   @Test
   fun `getAllLocations should return No data found when no location exists`() {
-    Mockito.`when`(locationService.getAllLocations()).thenReturn(listOf())
```
Comment (JazJax) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppselectronicmonitoringdataplatformapi/controller/LocationControllerTest.kt:38]: Why is this removed? Does this no longer get called by new-line 51? :
`val result = LocationController(locationService).getAllLocations()`

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppselectronicmonitoringdataplatformapi/controller/LocationControllerTest.kt [hunk 5]
```diff
@@ -89,7 +99,6 @@ class LocationControllerTest {
   @Test
   fun `getLocationsByDeviceWearerId should return no data found when no location exists`() {
     val deviceWearerId = "b537065a-094e-47eb-8fab-9698a9664d35"
-    Mockito.`when`(locationService.getLocationsByDeviceWearerId(deviceWearerId)).thenReturn(listOf())
```
Comment (JazJax) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppselectronicmonitoringdataplatformapi/controller/LocationControllerTest.kt:92]: Same as above comment

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppselectronicmonitoringdataplatformapi/controller/LocationControllerTest.kt [hunk 6]
```diff
@@ -320,7 +330,6 @@ class LocationControllerTest {
   @Test
   fun `getLocationsByDeviceId should return no data found when no location exists`() {
     val deviceId = "b537065a-094e-47eb-8fab-9698a9664d35"
-    Mockito.`when`(locationService.getLocationsByDeviceId(deviceId)).thenReturn(listOf())
```
Comment (JazJax) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppselectronicmonitoringdataplatformapi/controller/LocationControllerTest.kt:323]: As above

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppselectronicmonitoringdataplatformapi/controller/LocationControllerTest.kt [hunk 7]
```diff
@@ -532,4 +542,214 @@ class LocationControllerTest {
...
+    verify(locationService, times(1)).aggregateLocationsByDeviceIdAndTimeFrameAndDuration(
+      any<String>(),
+      any<Date>(),
+      any<Date>(),
+      any<Int>(),
+    )
+  }
+
+  @Test
+  fun `aggregateLocationsByDeviceIdAndTimeFrameAndDuration should return bad request when it does not receive valid IsoDateTime for startDate`() {
```
Comment (JazJax) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppselectronicmonitoringdataplatformapi/controller/LocationControllerTest.kt:705]: Does this do something more than the test on line 572? They both seem to do about the same thing.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppselectronicmonitoringdataplatformapi/controller/LocationControllerTest.kt [hunk 8]
```diff
@@ -532,4 +542,214 @@ class LocationControllerTest {
...
+    verify(locationService, times(0)).aggregateLocationsByDeviceIdAndTimeFrameAndDuration(
+      any<String>(),
+      any<Date>(),
+      any<Date>(),
+      any<Int>(),
+    )
+  }
+
+  @Test
+  fun `aggregateLocationsByDeviceIdAndTimeFrameAndDuration should return bad request when it does not receive valid IsoDateTime for endDate`() {
```
Comment (JazJax) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppselectronicmonitoringdataplatformapi/controller/LocationControllerTest.kt:731]: As above comment

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppselectronicmonitoringdataplatformapi/model/LocationAggregation.kt (hunk 1 above)
Comment (JazJax) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppselectronicmonitoringdataplatformapi/model/LocationAggregation.kt:12]: On second look, I don't know that this appears to be called anywhere - do we need it, or can we delete it (and the setValue function it uses on line 9)

**Review by JazJax — CHANGES_REQUESTED** [review]
Overall looks great with good testing, have a few comments/bits I'm note sure about if you could either amend or reply to comments to explain. :)

**Review by mandanakhademi — APPROVED** [review]
